File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict
def solution(genres, plays):
    answer = []
    n = len(genres)
    times = defaultdict(int)
    for i in range(n):
        times[genres[i]] += plays[i]
    
    q=defaultdict(list)
    for i, (g,p) in enumerate(zip(genres, plays)):
        q[g].append([i,p])
        
    
    t = sorted(q.items(), key = lambda x:(sum(p for _, p in x[1])), reverse = True)

    for genre, play in t:
        play.sort(key = lambda x:(-x[1], x[0]))
        answer.extend([i for i, _ in play[:2]])
    
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
class Solution {
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};
        List<Integer> lst = new ArrayList<>();
        Map<String, Integer> total = new HashMap<>();
        Map<String, List<int[]>> songs = new HashMap<>();
        for(int i=0;i<plays.length;i++){
            String g = genres[i];
            int play = plays[i];
            
            total.put(g, total.getOrDefault(g, 0)+play);
            songs.putIfAbsent(g, new ArrayList<>());
            songs.get(g).add(new int[]{i, play});
        }
        System.out.println(songs);
        List<String> sortedGenres = new ArrayList<>(total.keySet());
        System.out.println(sortedGenres);
        sortedGenres.sort((a,b)-> total.get(b)-total.get(a));
        
        List<Integer> ans = new ArrayList<>();
        for(String name: sortedGenres){
            List<int[]> song = songs.get(name);
            song.sort((a,b)-> {
                     if(b[1]==a[1]) return a[0]-b[0];
                     return b[1]-a[1];
            });
            for(int i=0; i<Math.min(2, song.size()); i++){
                ans.add(song.get(i)[0]);
            }
            
        }
        
        
        return ans.stream().mapToInt(i->i).toArray();
    }
}